### PR TITLE
define a macro for checking function calls on main thread

### PR DESCRIPTION
**Pull request by kevinushey**
Addresses https://github.com/rstudio/rstudio/issues/10040.

This PR bundles a few changes:

1. It refactors our main-thread checks into a set of helper macros, which should be a bit easier to use and will also signal more information in developer builds;
2. It also allows child processes to instruct their supervisors whether their callbacks can only be run on the main thread, and uses that for the existing classes whose callbacks need to interface with R.

### CODE REVIEW

File: src/cpp/session/modules/SessionReticulate.cpp [hunk 1]
```diff
@@ -48,24 +48,19 @@ void updateReticulatePython(bool forInit)
    if (!forInit && s_pythonInitialized)
       return;
 
+   ASSERT_MAIN_THREAD;
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionReticulate.cpp:51]: Here and elsewhere, some context we used to show in the warning message is now absent -- should we find a way to reintroduce that, or does the new message have enough context to fully remove the old one?

**Review by jmcphers — CHANGES_REQUESTED** [review]

File: src/cpp/session/modules/SessionReticulate.cpp (hunk 1 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionReticulate.cpp:51]: IMHO we have enough context here.

I do think that if a thread is trying to figure out the default version of Python we should define and use an alternate thread-safe implementation, but given that I can't recall the exact context in which this was being touched by a separate thread we might want to keep the main-thread check to be safe.

**Review by jmcphers — APPROVED** [review]

File: src/cpp/r/RExec.cpp [hunk 2]
```diff
@@ -126,21 +128,18 @@ enum EvalType {
...
    EvalDirect  // use Rf_eval directly
 };
+
 Error evaluateExpressionsUnsafe(SEXP expr,
                                 SEXP envir,
                                 SEXP* pSEXP,
                                 sexp::Protect* pProtect,
                                 EvalType evalType)
 {
+   ASSERT_MAIN_THREAD;
```
Comment (jeffvroom) [inline src/cpp/r/RExec.cpp:138]: If we are not on the main thread, I like that the ASSERT_MAIN_THREAD will include the logError but won't it now just go ahead and execute the expression, rather than returning an rCodeExecutionError?  In my testing, an attempt to actually execute R code off the main thread would longjmp that thread back into the R exec loop and that was not a fun debugging experience!  When we return an error, it tends to just abort the operation, like with that bug saving RMarkdown documents... the only side-effects now are the logError and now I feel like it would crash.

File: src/cpp/r/session/graphics/RGraphicsDevice.cpp [hunk 3]
```diff
@@ -760,8 +762,13 @@ const double kDefaultDevicePixelRatio = 1.0;
    
 Error initialize(
             const FilePath& graphicsPath,
-            const boost::function<bool(double*,double*)>& locatorFunction)
-{      
+            const boost::function<bool(double*, double*)>& locatorFunction)
+{
+   // initialize tracing
+   std::string traceGraphicsDevice = core::system::getenv("RSTUDIO_TRACE_GRAPHICS_DEVICE");
```
Comment (jeffvroom) [inline src/cpp/r/session/graphics/RGraphicsDevice.cpp:768]: I'm ok with this change, but it seems unrelated to the rest of this PR. Just wanted to mention it in case it snuck in accidentally.

**Review by jeffvroom — COMMENTED** [review]
Awesome changes, thanks for cleaning this all up!  Just that one issue about whether this will allow R code to try and run off the main thread... I think we need to return an error rather than continue on in that case.

File: src/cpp/r/RExec.cpp (hunk 2 above)
Comment (kevinushey) [reply src/cpp/r/RExec.cpp]: Thanks, you're right. I've revised the PR to allow for early returns in these contexts.

File: src/cpp/r/session/graphics/RGraphicsDevice.cpp (hunk 3 above)
Comment (kevinushey) [reply src/cpp/r/session/graphics/RGraphicsDevice.cpp:768]: I wanted to sneak it in just because it didn't feel worth having a whole new PR for dev-internal code.

**Review by jeffvroom — APPROVED** [review]
LGTM!
